sdks/python/apache_beam/typehints/typehints.py

```diff
@@ -411,17 +416,23 @@ def __eq__(self, other):
   def __repr__(self):
     return 'Any'
 
+  def __hash__(self):
+    return hash(id(self))
+
   def type_check(self, instance):
     pass
 
 
 class TypeVariable(AnyTypeConstraint):
 
+  def __init__(self, name):
+    self.name = name
+
   def __eq__(self, other):
     return type(self) == type(other) and self.name == other.name
 
-  def __init__(self, name):
-    self.name = name
+  def __hash__(self):
```

Comment:
This is again the same issue as mentioned above.

I actually think it might make more sense to change `__eq__` to be instance specific. If I specify `K = beam.typehints.TypeVariable('K')` in two separate places in my code base, I probably don't want them to 'overlap'.
However, this might break some existing pipelines. I therefore tried to use the equivalent of the Python 2 code.

What do you think?

Reply:
I reproduced the BEAM-3730 but could not easily interpret the failure mode. It is clear that equality relationship is not defined correctly here. I'd like to understand the failure first to understand what assumptions Beam codebase has about typehints. I don't have a good answer right now. For the purpose of this change, I think keeping `hash(id(self))` is ok for now since we won't make things more broken with that, but we should add a TODO(BEAM-3730): Fix the contract between `__hash__` and `eq`. I'll try to add more context to the issue later.

Reply:
Can you please add a `TODO(BEAM-3730)` comment here too?